File: backend/Models/Entities/Submission.cs
using OnlineClassroomManagement.Helper.Constants;

namespace OnlineClassroomManagement.Models.Entities
{
    public class Submission
    {
        public int Id { get; set; }
        public ClassMember SubmitBy { get; set; }
        public DateTime? SubmittedTime { get; set; }
        public string? Content { get; set; }
        public List<Document> SubmittedFiles { get; set; } = new();
        public SubmissionStatus Status { get; set; }
        public Grade? Grade { get; set; }
        public List<QuizAnswer> Answers { get; set; } = new();
        public AssignmentGroup? AssignmentGroup { get; set; }
    }
}

File: frontend/src/types/submission.ts
import { AttachmentResponse } from "./assignment";
import { AssignmentGroupResponse } from "./assignmentGroup";
import { GradeResponse } from "./grade";
import {
  CreateQuizAnswerRequest,
  QuizAnswer,
  QuizAnswerResponse,
} from "./quiz";

export interface CreateSubmissionRequest {
  content?: string;
  documentIds: number[];
  answers: CreateQuizAnswerRequest[];
}

export interface CreateSubmissionResponse {
  id: number;
  submitById: number;
  submittedTime?: Date;
  content?: string;
  submittedFiles: AttachmentResponse[];
  status: SubmissionStatus;
  answers: QuizAnswerResponse[];
  assignmentGroup: AssignmentGroupResponse | null;
}

export interface UpdateSubmissionRequest {
  id: number;
  content?: string;
  documentIds: number[];
}

export interface SubmissionResponse {
  id: number;
  submitById: number;
  submitByName?: string;
  submitByAvatarUrl?: string
  submitByEmail?: string;
  submittedTime?: Date;
  content?: string;
  submittedFiles: AttachmentResponse[];
  status: SubmissionStatus;
  grade?: GradeResponse;
  answers: QuizAnswer[];
  assignmentGroup: AssignmentGroupResponse | null;
}

export enum SubmissionStatus {
  NotSubmitted = 0,
  Submitted = 1,
  Graded = 2,
}

// Statistics for grading progress and scores of an assignment
export interface GradingStatistics {
  totalSubmissions: number;
  gradedSubmissions: number;
  ungradedSubmissions: number;
  averageScore: number;
  highestScore: number;
  lowestScore: number;
  gradingProgress: number; // percentage 0..100
}
